## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1.  <a href="#first">Part 1 of week 3</a>

2.  <a href="#second">Part 2 of week 3</a>

3.  <a href="#third">Part 3 of week 3</a>

    </font>
    </div>

## 1. Part 1 of week 3 <a class="anchor" id="first"></a>

## Instruction (copied from Coursera):

For this assignment, you will be required to explore and cluster the neighborhoods in Toronto.

- Start by creating a new Notebook for this assignment. (OK!)

- Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas  dataframe like the one shown below:
<img src="https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/7JXaz3NNEeiMwApe4i-fLg_40e690ae0e927abda2d4bde7d94ed133_Screen-Shot-2018-06-18-at-7.17.57-PM.png?expiry=1614556800000&hmac=wEwibcmT8gvyPaPIsVVZleNL3OOUvQYRv4pItozNzn4"/>

- To create the above dataframe:
    - The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood __(OK! Step "c")__
    - Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned. __(OK! Step "e")__
    - More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table.__(OK! Not necessary, as comented in step "f".)__
    - If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough. __(OK! Step "g")__
    - Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making. __(OK!)__
    - In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe. __(OK! Step "h")__

- Submit a link to your Notebook on your Github repository.(10 marks) __(OK!)__

## a) Import the necessary libraries

In [1]:
import pandas as pd
import numpy as np

## b) Read the tables at the wikipedia url

In [2]:
url = 'http://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
df_wiki1 = pd.read_html(url)
df_wiki1[0]

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

## c) Create a data frame with the table we are interested

In [3]:
df_wiki2 = df_wiki1[0]
df_wiki2

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

## d) Replace the "Not assigned" boroughs to "NaN"

In [4]:
df_wiki2['Borough'].replace('Not assigned',np.nan, inplace=True)
df_wiki2

Postal Code           Borough  \
0           M1A               NaN   
1           M2A               NaN   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z               NaN   
176         M6Z               NaN   
177         M7Z               NaN   
178         M8Z         Etobicoke   
179         M9Z               NaN   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

## e) Drop the rows where the borough was 'Not assigned'

In [5]:
df_wiki2.dropna(axis=0,inplace=True)
df_wiki2

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## f) Verify if there is a repeated 'Postal Code'

In [6]:
df_wiki2.nunique()

Postal Code      103
Borough           10
Neighbourhood     99
dtype: int64

## Analysis:
#### There are 103 unique Postal Codes and 103 rows at the table. 
#### Thus, it's not necessary to do this step:
_"More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table."_

## g) Replace the Neighbourhood with 'Not assigned' with the name of the corresponding borough.

In [7]:
for i in range(df_wiki2.shape[0]):
    if df_wiki2.iloc[i][2] == 'Not assigned':
        df_wiki2.iloc[i][2] == df_wiki2.iloc[i][1]
df_wiki2        

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## h) In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe. 

In [8]:
df_wiki2.shape

(103, 3)

## i) Reset index

In [9]:
df_wiki2.reset_index(drop=True, inplace=True)
df_wiki2

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

## 2. Part 2 of week 3 <a class="anchor" id="second"></a>

## j) Installations

In [10]:
!conda install -c conda-forge geopy --yes

Solving environment: done

# All requested packages already installed.



In [11]:
import json # library to handle JSON files

In [12]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [13]:
import requests # library to handle requests

In [14]:
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [15]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

In [16]:
!conda install -c conda-forge folium=0.5.0 --yes 

Solving environment: done

# All requested packages already installed.



In [17]:
import folium # map rendering library

In [18]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [19]:
import geocoder # import geocoder

## k) Try to get latitude and longitude for postal code 'M5G' using Geocoder Python package

In [20]:
# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format('M5G'))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

#Didn't work!

KeyboardInterrupt: 

## l) I couldn't get latitude and longitude using Geocoder Python package. Thus, I'll use the csv file (that I saved at my GitHub).

In [21]:
df_lat_lng = pd.read_csv('https://raw.githubusercontent.com/lucianaguerra/Coursera_Capstone/Week-3-assignment/Geospatial_Coordinates.csv')
df_lat_lng

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

## m) Search postal code at 'df_lat_lng' and get latitude and longitude

In [22]:
for i in range(df_wiki2.shape[0]):
    for j in range(df_lat_lng.shape[0]):
        if df_wiki2.iloc[i][0] == df_lat_lng.iloc[j][0]:
            df_wiki2.at[i,'Latitude'] = df_lat_lng.iloc[j][1]
            df_wiki2.at[i,'Longitude'] = df_lat_lng.iloc[j][2]
df_wiki2

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

## 3. Part 3 of week 3 <a class="anchor" id="third"></a>

## Instruction (copied from Coursera): 

- Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you. 

- Just make sure:
    - to add enough Markdown cells to explain what you decided to do and to report any observations you make. 
    - to generate maps to visualize your neighborhoods and how they cluster together. 
    - Once you are happy with your analysis, submit a link to the new Notebook on your Github repository. (3 marks)

In [23]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_wiki2['Borough'].unique()),
        df_wiki2.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


## n) Use geopy library to get the latitude and longitude values of Toronto.

In [24]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


## o) Create a map of Toronto with neighborhoods superimposed on top.

In [25]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_wiki2['Latitude'], df_wiki2['Longitude'], df_wiki2['Borough'], df_wiki2['Neighborhood']):
    label = '{}, {}'.format(df_wiki2, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

KeyError: 'Neighborhood'

### Didn't work... I think it's because some 'Neighborhood' has more than one 'Neighborhoods' separeted with commas

 ## p) let's simplify...  create a new data frame without the 'Neighborhood' that has more than one 'Neighborhood'

In [64]:
toronto_data = df_wiki2[~df_wiki2['Neighbourhood'].str.contains(',', regex=False)]
toronto_data

Postal Code           Borough                          Neighbourhood  \
0          M3A        North York                              Parkwoods   
1          M4A        North York                       Victoria Village   
7          M3B        North York                              Don Mills   
10         M6B        North York                              Glencairn   
13         M3C        North York                              Don Mills   
14         M4C         East York                       Woodbine Heights   
15         M5C  Downtown Toronto                         St. James Town   
16         M6C              York                     Humewood-Cedarvale   
19         M4E      East Toronto                            The Beaches   
20         M5E  Downtown Toronto                            Berczy Park   
21         M6E              York                    Caledonia-Fairbanks   
22         M1G       Scarborough                                 Woburn   
23         M4G         East York                                Leaside   
24         M5G  Downtown Toronto                     Central Bay Street   
25         M6G  Downtown Toronto                               Christie   
26         M1H       Scarborough                              Cedarbrae   
27         M2H        North York                      Hillcrest Village   
29         M4H         East York                       Thorncliffe Park   
32         M1J       Scarborough                    Scarborough Village   
39         M2K        North York                        Bayview Village   
40         M3K        North York                              Downsview   
46         M3L        North York                              Downsview   
50         M9L        North York                          Humber Summit   
53         M3M        North York                              Downsview   
54         M4M      East Toronto                        Studio District   
60         M3N        North York                              Downsview   
61         M4N   Central Toronto                          Lawrence Park   
62         M5N   Central Toronto                               Roselawn   
64         M9N              York                                 Weston   
66         M2P        North York                        York Mills West   
67         M4P   Central Toronto                       Davisville North   
70         M9P         Etobicoke                              Westmount   
76         M7R       Mississauga  Canada Post Gateway Processing Centre   
78         M1S       Scarborough                              Agincourt   
79         M4S   Central Toronto                             Davisville   
91         M4W  Downtown Toronto                               Rosedale   
92         M5W  Downtown Toronto                         Stn A PO Boxes   
95         M1X       Scarborough                            Upper Rouge   
99         M4Y  Downtown Toronto                   Church and Wellesley   

     Latitude  Longitude  
0   43.753259 -79.329656  
1   43.725882 -79.315572  
7   43.745906 -79.352188  
10  43.709577 -79.445073  
13  43.725900 -79.340923  
14  43.695344 -79.318389  
15  43.651494 -79.375418  
16  43.693781 -79.428191  
19  43.676357 -79.293031  
20  43.644771 -79.373306  
21  43.689026 -79.453512  
22  43.770992 -79.216917  
23  43.709060 -79.363452  
24  43.657952 -79.387383  
25  43.669542 -79.422564  
26  43.773136 -79.239476  
27  43.803762 -79.363452  
29  43.705369 -79.349372  
32  43.744734 -79.239476  
39  43.786947 -79.385975  
40  43.737473 -79.464763  
46  43.739015 -79.506944  
50  43.756303 -79.565963  
53  43.728496 -79.495697  
54  43.659526 -79.340923  
60  43.761631 -79.520999  
61  43.728020 -79.388790  
62  43.711695 -79.416936  
64  43.706876 -79.518188  
66  43.752758 -79.400049  
67  43.712751 -79.390197  
70  43.696319 -79.532242  
76  43.636966 -79.615819  
78  43.794200 -79.262029  
79  43.704324 -79.388790  
91  43.679563 -79.377529  
92 

## q) Create a new dataframe with only the neighborhoods in 'North York'

In [66]:
north_york = toronto_data[toronto_data['Borough'] == 'North York'].reset_index(drop=False)
north_york

index Postal Code     Borough      Neighbourhood   Latitude  Longitude
0       0         M3A  North York          Parkwoods  43.753259 -79.329656
1       1         M4A  North York   Victoria Village  43.725882 -79.315572
2       7         M3B  North York          Don Mills  43.745906 -79.352188
3      10         M6B  North York          Glencairn  43.709577 -79.445073
4      13         M3C  North York          Don Mills  43.725900 -79.340923
5      27         M2H  North York  Hillcrest Village  43.803762 -79.363452
6      39         M2K  North York    Bayview Village  43.786947 -79.385975
7      40         M3K  North York          Downsview  43.737473 -79.464763
8      46         M3L  North York          Downsview  43.739015 -79.506944
9      50         M9L  North York      Humber Summit  43.756303 -79.565963
10     53         M3M  North York          Downsview  43.728496 -79.495697
11     60         M3N  North York          Downsview  43.761631 -79.520999
12     66         M2P  North York    York Mills West  43.752758 -79.400049

## r) Let's get the geographical coordinates of North York.

In [67]:
address = 'North York, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of North York are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of North York are 43.7543263, -79.44911696639593.


## s) Let's visualizate North York the neighborhoods in it.

In [79]:
# create map of North York using latitude and longitude values
map_north_york = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(north_york['Latitude'], north_york['Longitude'], north_york['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_north_york)  
    
map_north_york

## t) Define Foursquare Credentials and Version

In [80]:
CLIENT_ID = 'EUTNZZ0U0ESN0SLHZO0DMXYCMWLKK3AW3R5JUBOY2J4LH2DM' # your Foursquare ID
CLIENT_SECRET = 'CXYVFP10HYPPAX44ASSOCMWZHDMANQORFUMPQITHE01RFUIW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: EUTNZZ0U0ESN0SLHZO0DMXYCMWLKK3AW3R5JUBOY2J4LH2DM
CLIENT_SECRET:CXYVFP10HYPPAX44ASSOCMWZHDMANQORFUMPQITHE01RFUIW


## u) Let's explore the first neighborhood in our dataframe.


In [84]:
north_york.iloc[0,3]

'Parkwoods'

## v) Get the neighborhood's latitude and longitude values

In [86]:
neighborhood_latitude = north_york.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = north_york.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = north_york.iloc[0, 3] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


## w) Now, let's get the top 100 venues that are in Parkwoods within a radius of 500 meters.

In [87]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=EUTNZZ0U0ESN0SLHZO0DMXYCMWLKK3AW3R5JUBOY2J4LH2DM&client_secret=CXYVFP10HYPPAX44ASSOCMWZHDMANQORFUMPQITHE01RFUIW&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

## x) Send the GET request and examine the resutls

In [88]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '603bdcc8f53fc41dc4abeebf'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 3,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

## y) Let's borrow the get_category_type function from the Foursquare lab.

In [89]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

## z) Clean the json and structure it into a pandas dataframe.

In [90]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1  Brookbanks Pool               Pool  43.751389 -79.332184
2    Variety Store  Food & Drink Shop  43.751974 -79.333114

## aa) How many venues were returned by Foursquare?

In [91]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

3 venues were returned by Foursquare.


## ab) Let's create a function to repeat the same process to all the neighborhoods in North York

In [92]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

## ac) Let's write the code to run the above function on each neighborhood and create a new dataframe called north_york_venues.

In [116]:
north_york_venues = getNearbyVenues(names=north_york['Neighbourhood'],
                                   latitudes=north_york['Latitude'],
                                   longitudes=north_york['Longitude']
                                  )


Parkwoods
Victoria Village
Don Mills
Glencairn
Don Mills
Hillcrest Village
Bayview Village
Downsview
Downsview
Humber Summit
Downsview
Downsview
York Mills West


## ad) Let's check the size of the resulting dataframe

In [117]:
print(north_york_venues.shape)
north_york_venues.head()

(68, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1         Brookbanks Pool       43.751389       -79.332184   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1                   Pool  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

## ae) Let's check how many venues were returned for each neighborhood

In [118]:
north_york_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                              
Bayview Village                        4                       4      4   
Don Mills                             26                      26     26   
Downsview                             17                      17     17   
Glencairn                              4                       4      4   
Hillcrest Village                      5                       5      5   
Humber Summit                          1                       1      1   
Parkwoods                              3                       3      3   
Victoria Village                       6                       6      6   
York Mills West                        2                       2      2   

                   Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                        
Bayview Village                 4                4               4  
Don Mills                      26               26              26  
Downsview                      17               17              17  
Glencairn                       4                4               4  
Hillcrest Village               5                5               5  
Humber Summit                   1                1               1  
Parkwoods                       3                3               3  
Victoria Village                6                6               6  
York Mills West                 2                2               2

## af) How many unique categories can be curated from all the returned venues

In [119]:
print('There are {} uniques categories.'.format(len(north_york_venues['Venue Category'].unique())))

There are 47 uniques categories.


## ag) Let's analyze Each Neighborhood

In [122]:
# one hot encoding
north_york_onehot = pd.get_dummies(north_york_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
north_york_onehot['Neighborhood'] = north_york_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [north_york_onehot.columns[-1]] + list(north_york_onehot.columns[:-1])
north_york_onehot = north_york_onehot[fixed_columns]

north_york_onehot.head()

Neighborhood  Airport  Art Gallery  Asian Restaurant  \
0         Parkwoods        0            0                 0   
1         Parkwoods        0            0                 0   
2         Parkwoods        0            0                 0   
3  Victoria Village        0            0                 0   
4  Victoria Village        0            0                 0   

   Athletics & Sports  Bakery  Bank  Baseball Field  Beer Store  Bike Shop  \
0                   0       0     0               0           0          0   
1                   0       0     0               0           0          0   
2                   0       0     0               0           0          0   
3                   0       0     0               0           0          0   
4                   0       0     0               0           0          0   

   ...  Pizza Place  Pool  Portuguese Restaurant  Pub  Restaurant  \
0  ...            0     0                      0    0           0   
1  ...            0     1                      0    0           0   
2  ...            0     0                      0    0           0   
3  ...            0     0                      0    0           0   
4  ...            0     0                      1    0           0   

   Sandwich Place  Shopping Mall  Sporting Goods Shop  Supermarket  \
0               0              0                    0            0   
1               0              0                    0            0   
2               0              0                    0            0   
3               0              0                    0            0   
4               0              0                    0            0   

   Sushi Restaurant  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 48 columns]

## ah) The new dataframe size.

In [124]:
north_york_onehot.shape

(68, 48)

## ai) let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [125]:
north_york_grouped = north_york_onehot.groupby('Neighborhood').mean().reset_index()
north_york_grouped

Neighborhood   Airport  Art Gallery  Asian Restaurant  \
0    Bayview Village  0.000000     0.000000          0.000000   
1          Don Mills  0.000000     0.038462          0.038462   
2          Downsview  0.058824     0.000000          0.000000   
3          Glencairn  0.000000     0.000000          0.000000   
4  Hillcrest Village  0.000000     0.000000          0.000000   
5      Humber Summit  0.000000     0.000000          0.000000   
6          Parkwoods  0.000000     0.000000          0.000000   
7   Victoria Village  0.000000     0.000000          0.000000   
8    York Mills West  0.000000     0.000000          0.000000   

   Athletics & Sports  Bakery      Bank  Baseball Field  Beer Store  \
0            0.000000    0.00  0.250000        0.000000    0.000000   
1            0.000000    0.00  0.000000        0.000000    0.076923   
2            0.058824    0.00  0.058824        0.058824    0.000000   
3            0.000000    0.25  0.000000        0.000000    0.000000   
4            0.000000    0.00  0.000000        0.000000    0.000000   
5            0.000000    0.00  0.000000        0.000000    0.000000   
6            0.000000    0.00  0.000000        0.000000    0.000000   
7            0.000000    0.00  0.000000        0.000000    0.000000   
8            0.000000    0.00  0.000000        0.000000    0.000000   

   Bike Shop  ...  Pizza Place      Pool  Portuguese Restaurant   Pub  \
0   0.000000  ...     0.000000  0.000000               0.000000  0.00   
1   0.038462  ...     0.000000  0.000000               0.000000  0.00   
2   0.000000  ...     0.000000  0.000000               0.000000  0.00   
3   0.000000  ...     0.250000  0.000000               0.000000  0.25   
4   0.000000  ...     0.000000  0.200000               0.000000  0.00   
5   0.000000  ...     1.000000  0.000000               0.000000  0.00   
6   0.000000  ...     0.000000  0.333333               0.000000  0.00   
7   0.000000  ...     0.166667  0.000000               0.166667  0.00   
8   0.000000  ...     0.000000  0.000000               0.000000  0.00   

   Restaurant  Sandwich Place  Shopping Mall  Sporting Goods Shop  \
0    0.000000        0.000000       0.000000             0.000000   
1    0.076923        0.038462       0.038462             0.038462   
2    0.000000        0.000000       0.058824             0.000000   
3    0.000000        0.000000       0.000000             0.000000   
4    0.000000        0.000000       0.000000             0.000000   
5    0.000000        0.000000       0.000000             0.000000   
6    0.000000        0.000000       0.000000             0.000000   
7    0.000000        0.000000       0.000000             0.000000   
8    0.000000        0.000000       0.000000             0.000000   

   Supermarket  Sushi Restaurant  
0     0.000000          0.000000  
1     0.038462          0.038462  
2     0.000000          0.000000  
3     0.000000          0.000000  
4     0.000000          0.000000  
5     0.000000          0.000000  
6     0.000000          0.000000  
7     0.000000          0.000000  
8     0.000000          0.000000  

[9 rows x 48 columns]

## aj) Let's confirm the new size

In [126]:
north_york_grouped.shape

(9, 48)

## ak) Let's print each neighborhood along with the top 5 most common venues

In [127]:
num_top_venues = 5

for hood in north_york_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = north_york_grouped[north_york_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Bayview Village----
                 venue  freq
0                 Bank  0.25
1  Japanese Restaurant  0.25
2                 Café  0.25
3   Chinese Restaurant  0.25
4              Airport  0.00


----Don Mills----
                  venue  freq
0                   Gym  0.12
1           Coffee Shop  0.08
2            Beer Store  0.08
3            Restaurant  0.08
4  Caribbean Restaurant  0.04


----Downsview----
                  venue  freq
0         Grocery Store  0.18
1                  Park  0.12
2               Airport  0.06
3          Liquor Store  0.06
4  Gym / Fitness Center  0.06


----Glencairn----
                 venue  freq
0               Bakery  0.25
1                  Pub  0.25
2          Pizza Place  0.25
3  Japanese Restaurant  0.25
4              Airport  0.00


----Hillcrest Village----
                      venue  freq
0  Mediterranean Restaurant   0.2
1                      Pool   0.2
2      Fast Food Restaurant   0.2
3                   Dog Run   0.2
4         

## al) Let's put that into a pandas dataframe

### First, let's write a function to sort the venues in descending order.

In [128]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [149]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = north_york_grouped['Neighborhood']

for ind in np.arange(north_york_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(north_york_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood     1st Most Common Venue 2nd Most Common Venue  \
0    Bayview Village                      Bank   Japanese Restaurant   
1          Don Mills                       Gym           Coffee Shop   
2          Downsview             Grocery Store                  Park   
3          Glencairn                    Bakery                   Pub   
4  Hillcrest Village  Mediterranean Restaurant                  Pool   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Café    Chinese Restaurant               Airport   
1            Beer Store            Restaurant  Caribbean Restaurant   
2               Airport          Liquor Store  Gym / Fitness Center   
3           Pizza Place   Japanese Restaurant               Airport   
4  Fast Food Restaurant               Dog Run           Golf Course   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                  Park  Gym / Fitness Center          Hockey Arena   
1    Italian Restaurant   Japanese Restaurant           Art Gallery   
2      Business Service            Food Truck                 Hotel   
3          Liquor Store                   Gym  Gym / Fitness Center   
4        Sandwich Place            Restaurant                   Gym   

  9th Most Common Venue 10th Most Common Venue  
0          Home Service                  Hotel  
1        Discount Store     Dim Sum Restaurant  
2         Shopping Mall         Baseball Field  
3          Hockey Arena           Home Service  
4  Gym / Fitness Center           Hockey Arena

 # Let's Cluster Neighborhoods

## am) Run k-means to cluster the neighborhood into 5 clusters.

In [150]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [151]:
# set number of clusters
kclusters = 5

north_york_grouped_clustering = north_york_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(north_york_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 1, 0, 3, 2, 4], dtype=int32)

## an) Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [152]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [158]:
north_york_merged = north_york
north_york_merged.rename(columns={'Neighbourhood': 'Neighborhood'}, inplace = True) # I don't know, but there was a problem of spellin 'neighborhood' and 'neighbourhood'
north_york_merged

index Postal Code     Borough       Neighborhood   Latitude  Longitude
0       0         M3A  North York          Parkwoods  43.753259 -79.329656
1       1         M4A  North York   Victoria Village  43.725882 -79.315572
2       7         M3B  North York          Don Mills  43.745906 -79.352188
3      10         M6B  North York          Glencairn  43.709577 -79.445073
4      13         M3C  North York          Don Mills  43.725900 -79.340923
5      27         M2H  North York  Hillcrest Village  43.803762 -79.363452
6      39         M2K  North York    Bayview Village  43.786947 -79.385975
7      40         M3K  North York          Downsview  43.737473 -79.464763
8      46         M3L  North York          Downsview  43.739015 -79.506944
9      50         M9L  North York      Humber Summit  43.756303 -79.565963
10     53         M3M  North York          Downsview  43.728496 -79.495697
11     60         M3N  North York          Downsview  43.761631 -79.520999
12     66         M2P  North York    York Mills West  43.752758 -79.400049

In [159]:
# merge north_york_grouped with north_york to add latitude/longitude for each neighborhood
north_york_merged = north_york_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

In [160]:
north_york_merged.head() 

index Postal Code     Borough      Neighborhood   Latitude  Longitude  \
0      0         M3A  North York         Parkwoods  43.753259 -79.329656   
1      1         M4A  North York  Victoria Village  43.725882 -79.315572   
2      7         M3B  North York         Don Mills  43.745906 -79.352188   
3     10         M6B  North York         Glencairn  43.709577 -79.445073   
4     13         M3C  North York         Don Mills  43.725900 -79.340923   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3     Food & Drink Shop                  Pool   
1               2     French Restaurant          Hockey Arena   
2               2                   Gym           Coffee Shop   
3               2                Bakery                   Pub   
4               2                   Gym           Coffee Shop   

  3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0                  Park                Airport          Liquor Store   
1          Intersection  Portuguese Restaurant           Pizza Place   
2            Beer Store             Restaurant  Caribbean Restaurant   
3           Pizza Place    Japanese Restaurant               Airport   
4            Beer Store             Restaurant  Caribbean Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                   Gym  Gym / Fitness Center          Hockey Arena   
1           Coffee Shop          Liquor Store                   Gym   
2    Italian Restaurant   Japanese Restaurant           Art Gallery   
3          Liquor Store                   Gym  Gym / Fitness Center   
4    Italian Restaurant   Japanese Restaurant           Art Gallery   

  9th Most Common Venue 10th Most Common Venue  
0          Home Service                  Hotel  
1  Gym / Fitness Center           Home Service  
2        Discount Store     Dim Sum Restaurant  
3          Hockey Arena           Home Service  
4        Discount Store     Dim Sum Restaurant

## ao) Let's visualize the resulting clusters

In [162]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(north_york_merged['Latitude'], north_york_merged['Longitude'], north_york_merged['Neighborhood'], north_york_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Let's Examine Clusters


## Cluster 1

In [163]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 0, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
9         M9L -79.565963               0           Pizza Place   

  2nd Most Common Venue     3rd Most Common Venue 4th Most Common Venue  \
9               Airport  Mediterranean Restaurant                   Gym   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9  Gym / Fitness Center          Hockey Arena          Home Service   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
9                 Hotel          Intersection     Italian Restaurant

## Cluster 2

In [164]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 1, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels     1st Most Common Venue  \
5         M2H -79.363452               1  Mediterranean Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5                  Pool  Fast Food Restaurant               Dog Run   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5           Golf Course        Sandwich Place            Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5                   Gym  Gym / Fitness Center           Hockey Arena

## Cluster 3

In [165]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 2, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
1          M4A -79.315572               2     French Restaurant   
2          M3B -79.352188               2                   Gym   
3          M6B -79.445073               2                Bakery   
4          M3C -79.340923               2                   Gym   
6          M2K -79.385975               2                  Bank   
7          M3K -79.464763               2         Grocery Store   
8          M3L -79.506944               2         Grocery Store   
10         M3M -79.495697               2         Grocery Store   
11         M3N -79.520999               2         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1           Hockey Arena          Intersection  Portuguese Restaurant   
2            Coffee Shop            Beer Store             Restaurant   
3                    Pub           Pizza Place    Japanese Restaurant   
4            Coffee Shop            Beer Store             Restaurant   
6    Japanese Restaurant                  Café     Chinese Restaurant   
7                   Park               Airport           Liquor Store   
8                   Park               Airport           Liquor Store   
10                  Park               Airport           Liquor Store   
11                  Park               Airport           Liquor Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1            Pizza Place           Coffee Shop          Liquor Store   
2   Caribbean Restaurant    Italian Restaurant   Japanese Restaurant   
3                Airport          Liquor Store                   Gym   
4   Caribbean Restaurant    Italian Restaurant   Japanese Restaurant   
6                Airport                  Park  Gym / Fitness Center   
7   Gym / Fitness Center      Business Service            Food Truck   
8   Gym / Fitness Center      Business Service            Food Truck   
10  Gym / Fitness Center      Business Service            Food Truck   
11  Gym / Fitness Center      Business Service            Food Truck   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                    Gym  Gym / Fitness Center           Home Service  
2            Art Gallery        Discount Store     Dim Sum Restaurant  
3   Gym / Fitness Center          Hockey Arena           Home Service  
4            Art Gallery        Discount Store     Dim Sum Restaurant  
6           Hockey Arena          Home Service                  Hotel  
7                  Hotel         Shopping Mall         Baseball Field  
8                  Hotel         Shopping Mall         Baseball Field  
10                 Hotel         Shopping Mall         Baseball Field  
11                 Hotel         Shopping Mall         Baseball Field

## Cluster 4

In [166]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 3, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
0         M3A -79.329656               3     Food & Drink Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Pool                  Park               Airport   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Liquor Store                   Gym  Gym / Fitness Center   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Hockey Arena          Home Service                  Hotel

## Cluster 5

In [167]:
north_york_merged.loc[north_york_merged['Cluster Labels'] == 4, north_york_merged.columns[[1] + list(range(5, north_york_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
12         M2P -79.400049               4     Convenience Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12                  Park               Airport          Liquor Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12                   Gym  Gym / Fitness Center          Hockey Arena   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12          Home Service                 Hotel           Intersection